### PR TITLE
Use pagination when requesting github API

### PR DESCRIPTION
**Pull request by mboukhalfa**
Fix https://github.com/metal3-io/metal3-dev-env/issues/1162

### CODE REVIEW

File: lib/releases.sh [hunk 1]
```diff
@@ -2,29 +2,49 @@
 
 function get_latest_release() {
   set +x
+  # get last page
   if [ -z "${GITHUB_TOKEN:-}" ]; then
-    release="$(curl -sL "${1}")" || ( set -x && exit 1 )
+    last_page=$(curl -s -I "${1}" -H "Authorization: token $GITHUB_TOKEN" | grep '^link:' | sed -e 's/^link:.*page=//g' -e 's/>.*$//g') || (set -x && exit 1)
```
Comment (tuminoid) [inline lib/releases.sh:7]: `(set -x && exit 1)` creates a subshell, the parenthesis does that, so the `exit 1` exits only that subshell, and not the script. If you want to perform two actions, use curly braces, `{set -x && exit 1}`.

Also, you want to assign the subshell output of `$( ... )` with quotes, so do `last_page="$( ... )"`

File: lib/releases.sh [hunk 2]
```diff
@@ -2,29 +2,49 @@
...
   else
-    release="$(curl -H "Authorization: token ${GITHUB_TOKEN}" -sL "${1}")" || ( set -x && exit 1 )
+    last_page=$(curl -s -I "${1}" | grep '^link:' | sed -e 's/^link:.*page=//g' -e 's/>.*$//g') || (set -x && exit 1)
```
Comment (tuminoid) [inline lib/releases.sh:9]: same here

File: lib/releases.sh [hunk 3]
```diff
@@ -2,29 +2,49 @@
...
   fi
-  # This gets the latest release as vx.y.z or vx.y.z-rc.0, including any version with a suffix starting with - , for example -rc.0
-  # The order is exactly as released in Github.
-  # Downside is that selecting official releases only isn't possible, while pre-release
-  # selection is possible given specific enough prefix, like v1.3.0-pre
-  release_tag="$(echo "$release" | jq -r "[.[].tag_name | select( startswith(\"${2:-}\"))] | .[0]")"
+  # default last page to 1
+  last_page=${last_page:-1}
```
Comment (tuminoid) [inline lib/releases.sh:12]: Quotes do no harm here either.

File: lib/releases.sh [hunk 4]
```diff
@@ -2,29 +2,49 @@
...
 
+  for current_page in $(seq 1 "$last_page"); do
```
Comment (tuminoid) [inline lib/releases.sh:14]: Any time you want to use `seq` to create sequences in Bash for loops, use `{1..$last_page}` instead. This avoids various issues with subshells, output etc.

File: lib/releases.sh [hunk 5]
```diff
@@ -2,29 +2,49 @@
...
+
+    url="${1}?page=${current_page}"
+
+    if [ -z "${GITHUB_TOKEN:-}" ]; then
+      release="$(curl -sL "${url}")" || (set -x && exit 1)
```
Comment (tuminoid) [inline lib/releases.sh:19]: same as the first one

File: lib/releases.sh [hunk 6]
```diff
@@ -2,29 +2,49 @@
...
+    else
+      release="$(curl -H "Authorization: token ${GITHUB_TOKEN}" -sL "${url}")" || (set -x && exit 1)
```
Comment (tuminoid) [inline lib/releases.sh:21]: same as the first one

File: lib/releases.sh [hunk 7]
```diff
@@ -2,29 +2,49 @@
...
+    fi
+
+    # This gets the latest release as vx.y.z or vx.y.z-rc.0, including any version with a suffix starting with - , for example -rc.0
+    # The order is exactly as released in Github.
+    # Downside is that selecting official releases only isn't possible, while pre-release
+    # selection is possible given specific enough prefix, like v1.3.0-pre
+    release_tag="$(echo "$release" | jq -r "[.[].tag_name | select( startswith(\"${2:-}\"))] | .[0]")"
```
Comment (tuminoid) [inline lib/releases.sh:28]: Add curlys to `$release`

File: lib/releases.sh [hunk 8]
```diff
@@ -2,29 +2,49 @@
...
+    # if release tag found
+    if [[ "$release_tag" != "null" ]]; then
```
Comment (tuminoid) [inline lib/releases.sh:30]: same

File: lib/releases.sh (hunk 4 above)
Comment (mboukhalfa) [reply lib/releases.sh]: done

Comment (mboukhalfa) [reply lib/releases.sh]: (warning) Bash doesn't support variables in brace range expansions !

Comment (tuminoid) [reply lib/releases.sh]: Duh, my bad. See, Bash is hard :)

File: lib/releases.sh (hunk 1 above)
Comment (tuminoid) [reply lib/releases.sh]: `${set -x ...}` is definitely wrong, but what does it complain about `{ set -x ... }` ?

Comment (mboukhalfa) [reply lib/releases.sh]: thanks! done

**Comment by mboukhalfa** [comment]
/assign @furkatgofurov7
cc @lentzi90

File: lib/releases.sh [hunk 9]
```diff
@@ -49,13 +70,13 @@ BMOBRANCH="${BMOBRANCH:-${BMORELEASE}}"
...
 
 if [[ "$CAPM3RELEASE" == "" ]]; then
-  command -v jq &> /dev/null && echo "Failed to fetch CAPM3 release from Github" && exit 1
+  command -v jq &>/dev/null && echo "Failed to fetch CAPM3 release from Github" && exit 1
 fi
 
 if [[ "$BMORELEASE" == "" ]]; then
-  command -v jq &> /dev/null && echo "Failed to fetch BMO release from Github" && exit 1
+  command -v jq &>/dev/null && echo "Failed to fetch BMO release from Github" && exit 1
 fi
```
Comment (lentzi90) [inline lib/releases.sh:82]: Why remove the space here? I think it is easier to read with the space so we should add it back.

**Review by furkatgofurov7 — APPROVED** [review]
Checked locally, it works fine

/approve
/hold to address the comments from @lentzi90

File: lib/releases.sh (hunk 9 above)
Comment (mboukhalfa) [reply lib/releases.sh:82]: it was aut reformatted by editor linter what do we have coding style for that

Comment (lentzi90) [reply lib/releases.sh:82]: @tuminoid any preference on this? I'm inclined to approve it as just a matter of preference if there are no objections.

Comment (tuminoid) [reply lib/releases.sh:82]: That space is completely up to personal taste. I tend to have the space there for readability, but it doesn't really matter.

Comment (lentzi90) [reply lib/releases.sh:82]: Alright I'm putting lgtm then :slightly_smiling_face:

**Pull request merged**